### ISSUE TITLE
Implement Ascon

### ISSUE THREAD

**Issue by paulmillr** [paulmillr/noble-ciphers#12]
Ascon won NIST lightweight cryptography contest: https://csrc.nist.gov/News/2023/lightweight-cryptography-nist-selects-ascon

**Comment by mirceanis** [comment]
➕ 💯  for AESKW
 
that would allow me to greatly simplify the dependencies in several projects I maintain, where I have to use a very heavy isomorphic-webcrypto otherwise

**Comment by paulmillr** [comment]
@mirceanis AESKW is done.

**Comment by alterstep** [comment]
I would like AEGIS. There is a JavaScript implementation, but it's only AEGIS-128L and is not very good.

**Comment by paulmillr** [comment]
@alterstep any chance you'll be able to describe what are you using it for?

**Comment by alterstep** [comment]
> @alterstep any chance you'll be able to describe what are you using it for?

Yes, I use libsodium on the server because AEGIS is very fast. On the client side, I could use libsodium-js, but it's very large, and I don't need its full functionality. Noble-ciphers would be perfect.

**Comment by paulmillr** [comment]
Yes, but why do you need something "very fast"? Why can't AES work for you?

**Comment by alterstep** [comment]
> Yes, but why do you need something "very fast"? Why can't AES work for you?

256-bit nonces too.

**Comment by paulmillr** [comment]
Why AEGIS instead of XChacha20-poly1305?

**Comment by alterstep** [comment]
sodium AEGIS-256 is more than 5 times faster than XChaCha

**Comment by paulmillr** [comment]
Interesting, ok.

The question is, however, do you actually *need* this speed? Is the speed of algorithms a bottleneck in your application?

**Comment by BigPanda97** [comment]
Ascon and AEGIS are both nice. And there is an [Ascon-js](https://github.com/Simolation/ascon-js) version already available which could probably easily be implemented here.

**Comment by paulmillr** [comment]
Keep in mind it's not just "implementing" - it's also continuously supporting and perhaps having 3rd party audits of the code.

**Comment by paulmillr** [comment]
AEGIS has been implemented here: https://github.com/stknob/aegis-ts/tree/main

**Comment by BigPanda97** [comment]
I would still recommend implementing them, as Ascon and AEGIS both are very useful ciphers, probably much more than Salsa20.

**Comment by paulmillr** [comment]
@BigPanda97 what makes them more useful than Salsa for your use case?

**Comment by BigPanda97** [comment]
> @BigPanda97 what makes them more useful than Salsa for your use case?

This question seems more like a joke. Salsa is basically used nowhere nowadays except for Threema messenger which still uses it. However:

- AEGIS is currently being standardised as a new cipher suite for TLS (https://datatracker.ietf.org/doc/draft-denis-tls-aegis/) and is together with ChaCha20 the standard cipher of LibSodium (https://github.com/jedisct1/libsodium), the most used cryptographic library worldwide.

- ASCON is "the standard" by NIST for lightweight cryptography, and they published it now officially as standard "NIST SP 800-232" (https://csrc.nist.gov/pubs/sp/800/232/ipd) in November.

So both of them have and will have a very widespread usage just like AES, underlining their importance in every cryptographic library.

**Comment by paulmillr** [comment]
If you read topic start post, it mentions benefits already. There is no need to write it twice.

I'm not asking about benefits of algorithms. I'm asking how **you, personally** would benefit from having those algo / algos in noble. Details of your environment. Why can't it use ChaCha20poly1305.

> Salsa is basically used nowhere nowadays

[one](https://github.com/search?q=crypto_secretbox&type=code), [two](https://github.com/search?q=crypto_stream_xor&type=code)

**Comment by BigPanda97** [comment]
> If you read topic start post, it mentions benefits already. There is no need to write it twice.

Wrong, the implementation in TLS and the publishing of the final standardisation is new.

> I'm not asking about benefits of algorithms. I'm asking how **you, personally** would benefit from having those algo / algos in noble.

For example to decrypt stuff which has been encrypted using LibSodiums AEGIS? 🙄

> > Salsa is basically used nowhere nowadays

Your example searches reveal many old and outdated projects as well as many implementations where the underlying algorithms of the API have been switched to ChaCha instead. A productive usage of Salsa is useless nowadays, since ChaCha is both more secure and more efficient/faster.

There is still no valid reason, why any good and reputable cryptographic library should NOT implement AEGIS and ASCON. Both will be the algorithms of the future.

**Comment by paulmillr** [comment]
> There is still no valid reason, why any good and reputable cryptographic library should NOT implement AEGIS and ASCON

So submit a pull request.
